### PR TITLE
Move AMM into a package that collects RUN protocol components

### PR DESCRIPTION
**Pull request by Chris-Hibbert**
refs: #4199

## Description

Move the AMM out of the Zoe contacts directory and into packages/run-protocol.

### Security Considerations

This should be without security implications. It's just rearranging packages.

### Documentation Considerations

As mentioned in #4199 and #4209, the docs need to be reorganized to make RUN protocol more visible.

### Testing Considerations

Rearranging the test subdirectory to distinguish the various RUN protocol components.

### Updates to Type Declarations

As a result of the rearrangement, eslint reported some type inconsistencies that weren't previously evident. I suspect that type checking is loose in .../zoe/src/contracts or .../zoe/tests, but tighter in run-protocol.  In any case, I cleaned up some type declarations in .../src/.  I also added some `// @ts-ignore` comments in tests that were constructing parameter lists on the fly by destructuring objects.

### CODE REVIEW

File: packages/run-protocol/src/vpool-xyk-amm/types.js [hunk 1]
```diff
@@ -62,6 +62,8 @@
  * @property {(brand: Brand) => {toCentral: PriceAuthority, fromCentral: PriceAuthority}} getPriceAuthorities
  * get a pair of PriceAuthorities { toCentral, fromCentral } for requesting
  * Prices and notifications about changing prices.
+ * @property {() => Brand[]} getAllPoolBrands
+ * @property {() => Allocation} getProtocolPoolBalance
```
Comment (dckc) [inline packages/run-protocol/src/vpool-xyk-amm/types.js:66]: I'm curious what prompted this. It looks like a bit of a (welcome) drive-by.

File: packages/run-protocol/test/amm/constantProduct/test-checkInvariants.js [hunk 2]
```diff
@@ -2,15 +2,24 @@
...
+import { getXY } from '../../../src/vpool-xyk-amm/constantProduct/getXY.js';
+
+/**
+ * @typedef {Object} SwapPriceArgs
+ * @property {Amount} amountGiven
+ * @property {PoolAllocation} poolAllocation
+ * @property {Amount=} amountWanted
+ * @property {Ratio} protocolFeeRatio
+ * @property {Ratio} poolFeeRatio
+ */
```
Comment (dckc) [inline packages/run-protocol/test/amm/constantProduct/test-checkInvariants.js:22]: more (welcome) type tweaks... seems worth a mention in the commit message.

File: packages/run-protocol/test/amm/constantProduct/test-checkInvariants.js [hunk 3]
```diff
@@ -110,6 +119,7 @@ function checkGetOutput(t, args, result) {
 
 const testGetInputPrice = (t, inputs, runIn) => {
   const args = runIn ? prepareRUNInTest(inputs) : prepareRUNOutTest(inputs);
+  // @ts-ignore
```
Comment (dckc) [inline packages/run-protocol/test/amm/constantProduct/test-checkInvariants.js]: add a few words about why `@ts-ignore` is appropriate here?

File: packages/run-protocol/test/amm/constantProduct/test-compareBondingCurves.js [hunk 4]
```diff
@@ -39,6 +39,7 @@ const prepareSwapInTest = ({ inputReserve, outputReserve, inputValue }) => {
 
 const testInputGetPrice = (t, inputs, expectedOutput) => {
   const { args, bld } = prepareSwapInTest(inputs);
+  // @ts-ignore
```
Comment (dckc) [inline packages/run-protocol/test/amm/constantProduct/test-compareBondingCurves.js]: likewise for the other `@ts-ignore` occurrences

File: packages/run-protocol/test/vaultFactory/test-stablecoin.js [hunk 5]
```diff
@@ -18,11 +18,11 @@ import {
 import { makePromiseKit } from '@agoric/promise-kit';
 import { makeScriptedPriceAuthority } from '@agoric/zoe/tools/scriptedPriceAuthority.js';
 import { assertAmountsEqual } from '@agoric/zoe/test/zoeTestHelpers.js';
+import { makeParamManagerBuilder } from '@agoric/governance';
```
Comment (dckc) [inline packages/run-protocol/test/vaultFactory/test-stablecoin.js:21]: the name of this file should change, yes?

**Review by dckc — CHANGES_REQUESTED** [review]
renaming `test-stablecoin.js` seems worth doing

File: packages/run-protocol/src/vpool-xyk-amm/types.js (hunk 1 above)
Comment (Chris-Hibbert) [reply packages/run-protocol/src/vpool-xyk-amm/types.js:66]: I didn't investigate, but I suspected that moving tests around caused a change to the way declarations were imported or how eslint was configured in tests.  As I recall, we have eslint warnings turned off in some test directories.

File: packages/run-protocol/test/amm/constantProduct/test-checkInvariants.js (hunk 2 above)
Comment (Chris-Hibbert) [reply packages/run-protocol/test/amm/constantProduct/test-checkInvariants.js:22]: added a few lines to the message.

File: packages/run-protocol/test/amm/constantProduct/test-checkInvariants.js (hunk 3 above)
Comment (Chris-Hibbert) [reply packages/run-protocol/test/amm/constantProduct/test-checkInvariants.js]: done

File: packages/run-protocol/test/amm/constantProduct/test-compareBondingCurves.js (hunk 4 above)
Comment (Chris-Hibbert) [reply packages/run-protocol/test/amm/constantProduct/test-compareBondingCurves.js]: done

File: packages/run-protocol/test/vaultFactory/test-stablecoin.js (hunk 5 above)
Comment (Chris-Hibbert) [reply packages/run-protocol/test/vaultFactory/test-stablecoin.js:21]: good point.  I looked around and found a few files with treasury in their names and renamed them, too.

File: packages/run-protocol/test/amm/constantProduct/test-checkInvariants.js [hunk 6]
```diff
@@ -119,7 +119,7 @@ function checkGetOutput(t, args, result) {
 
 const testGetInputPrice = (t, inputs, runIn) => {
   const args = runIn ? prepareRUNInTest(inputs) : prepareRUNOutTest(inputs);
-  // @ts-ignore
+  // @ts-ignore typescript doesn't like param list built by destructuring
```
Comment (dckc) [inline packages/run-protocol/test/amm/constantProduct/test-checkInvariants.js:122]: I wonder if there's a way to use [`Parameters<Type>`](https://www.typescriptlang.org/docs/handbook/utility-types.html#parameterstype).

I'm not inspired to work it out just now, though.

**Review by dckc — APPROVED** [review]
